### PR TITLE
Object status filter on applications #1485

### LINKED ISSUE #1485: Application related configurations and `status` filter
* add `config.application_id` foreign key to load `application` specific configuration.
* handle `status` configuration with possible values:
  - `on` -> only **on** status is allowed
 - `draft` -> only **on** and **draft** status are allowed
 - `all` (default) -> all statuses are allowed

### PR DESCRIPTION
**Pull request by stefanorosanelli**
This PR fixes #1485

* a new `config.application_id` foreign key has been created
* upon application identification `core` context configuration for the selected application is loaded, via `application_id`
* `Staus.level`  configuration introduced with possibile values:
  - `'on'` => only objects with status `on` are loaded
  - `'draft'` => only objects with status `on` or `draft` are loaded
* `BaseAction`, `GetObjectAction` and `ListObjectsAction` have been modified. Are there other actions that need to be changed?
* a new integration test `StatusLevel` was created

### CODE REVIEW

File: plugins/BEdita/API/tests/IntegrationTest/StatusLevelTest.php [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        $this->assertResponseCode(200);
+        $result = json_decode((string)$this->_response->getBody(), true);
+
+        $ids = Hash::extract($result, 'data.{n}.id');
+        sort($ids);
+        static::assertEquals($expected, $ids);
+    }
+
+    /**
+     * Provider for testListObjects()
```
Comment (batopa) [inline plugins/BEdita/API/tests/IntegrationTest/StatusLevelTest.php]: typo: it should be `testSingleObject()`

File: plugins/BEdita/Core/src/Configure/Engine/DatabaseConfig.php [hunk 2]
```diff
@@ -40,6 +46,16 @@ class DatabaseConfig implements ConfigEngineInterface
      */
     protected $reservedKeys = ['Datasources', 'Cache', 'EmailTransport', 'Session', 'Error', 'App'];
 
+    /**
+     * Setup application id if
```
Comment (batopa) [inline plugins/BEdita/Core/src/Configure/Engine/DatabaseConfig.php]: `if` typo ?

File: plugins/BEdita/Core/src/Model/Action/BaseAction.php [hunk 3]
```diff
@@ -72,4 +73,23 @@ protected function initialize(array $config)
     {
         return $this->execute($data);
     }
+
+    /**
+     * Allowed object `status` condition
+     *
+     * @return array Empty array if all `status` are allowed otherwise a list of allowed values
+     */
+    protected function statusCondition()
```
Comment (batopa) [inline plugins/BEdita/Core/src/Model/Action/BaseAction.php]: I'm not sure this is the right place for that piece of logic. `BaseAction` is intended as base class for all actions not necessarily for BEdita Objects action, so  I think is preferable to move elsewhere, maybe in a Trait.

**Review by batopa — CHANGES_REQUESTED** [review]

File: plugins/BEdita/Core/src/Configure/Engine/DatabaseConfig.php (hunk 2 above)
Comment (stefanorosanelli) [reply plugins/BEdita/Core/src/Configure/Engine/DatabaseConfig.php]: if... I don't remember :smile_cat:

File: plugins/BEdita/Core/src/Model/Action/BaseAction.php (hunk 3 above)
Comment (stefanorosanelli) [reply plugins/BEdita/Core/src/Model/Action/BaseAction.php]: I wasn't sure about that either.
Could be a base class, but probably an `ObjectConditionsTrait` is better.

Comment (stefanorosanelli) [reply plugins/BEdita/Core/src/Model/Action/BaseAction.php]: done

File: plugins/BEdita/Core/src/Configure/Engine/DatabaseConfig.php (hunk 2 above)
Comment (stefanorosanelli) [reply plugins/BEdita/Core/src/Configure/Engine/DatabaseConfig.php]: fixed

File: plugins/BEdita/API/tests/IntegrationTest/StatusLevelTest.php (hunk 1 above)
Comment (stefanorosanelli) [reply plugins/BEdita/API/tests/IntegrationTest/StatusLevelTest.php]: fixed

File: plugins/BEdita/Core/src/Model/Action/BaseAction.php [hunk 4]
```diff
@@ -13,6 +13,7 @@
 
 namespace BEdita\Core\Model\Action;
 
+use Cake\Core\Configure;
```
Comment (batopa) [inline plugins/BEdita/Core/src/Model/Action/BaseAction.php]: unused `use` statement

**Review by batopa — APPROVED** [review]

File: plugins/BEdita/Core/src/Model/Action/BaseAction.php (hunk 4 above)
Comment (stefanorosanelli) [reply plugins/BEdita/Core/src/Model/Action/BaseAction.php]: done

File: plugins/BEdita/Core/src/Model/Action/GetObjectAction.php [hunk 5]
```diff
@@ -54,6 +56,7 @@ public function execute(array $data = [])
         $conditions = [
             'deleted' => (int)!empty($data['deleted']),
         ];
+        $conditions += $this->statusCondition();
```
Comment (fquffio) [inline plugins/BEdita/Core/src/Model/Action/GetObjectAction.php]: Couldn't we implement and use a custom finder in `ObjectsTable` instead? Finders are inherited from child tables, so it should be alright to use regardless of the actual table being used.

Something like:
```php
public function findStatus(Query $query, array $options)
{
    $mode = reset($options);
    switch ($mode) {
        case 'on':
            return $query->where([$this->aliasField('status') => 'on']);

        case 'draft':
            return $query
                ->where(function (QueryExpression $exp) {
                    return $exp->in($this->aliasField('status'), ['on', 'draft']);
                });

        case 'off':
        case 'all':
            return $query;

        default:
            throw new BadFilterException('Invalid mode');
    }
}
```

**Comment by fquffio** [comment]
After private discussions, I made the changes I had suggested in my review. The build is failing due to json-schema.org being currently unreachable: see #1490 for reference.

File: plugins/BEdita/Core/src/Model/Action/GetObjectAction.php [hunk 6]
```diff
@@ -51,21 +53,44 @@ protected function initialize(array $data)
...
+        if (count($key) !== count($primaryKey)) {
+            $primaryKey = $primaryKey ?: [null];
+            $primaryKey = array_map(function ($key) {
+                return var_export($key, true);
+            }, $primaryKey);
+
+            throw new InvalidPrimaryKeyException(sprintf(
+                'Record not found in table "%s" with primary key [%s]',
+                $this->Table->getTable(),
+                implode($primaryKey, ', ')
```
Comment (stefanorosanelli) [inline plugins/BEdita/Core/src/Model/Action/GetObjectAction.php]: Should this be `implode(', ', $primaryKey)` ?
http://php.net/manual/en/function.implode.php

Also: why this change?
Should we check for multi column primary keys on objects?

Comment (fquffio) [reply plugins/BEdita/Core/src/Model/Action/GetObjectAction.php]: Fixed. It would work anyway because [`implode()` accepts parameters in either order](http://php.net/implode), but it's definitely way way better how you suggested.

This snippet has been basically copy-pasted from CakePHP's `Table::get()`. I had to duplicate this portion of code because it is impossible to chain multiple finders with `Table::get()`, and as a consequence I had to perform some basic checks on primary key.

We don't need to support multi-column primary keys on objects, but we need to block attempts to pass multiple values for `primaryKey`:
```php
(new GetObjectAction(['table' => TableRegistry::get('Objects')]))
    ->execute(['primaryKey' => [1, 2, 3]]);
// This throws:
// new InvalidPrimaryKeyException(
//     'Record not found in table "objects" with primary key [1, 2, 3]'
// );
```

Comment (stefanorosanelli) [reply plugins/BEdita/Core/src/Model/Action/GetObjectAction.php]: Got it. Since this is a critical action I would perform this check in a dedicated method and/or add a longer comment to explain why we have to do this....

Comment (fquffio) [reply plugins/BEdita/Core/src/Model/Action/GetObjectAction.php]: Agreed.
